### ISSUE TITLE
Current issues with (meta) data upgrade

### ISSUE THREAD

**Issue by bleskes** [elastic/elasticsearch#13843]
Reviewing https://github.com/elastic/elasticsearch/pull/13799, I found it hard to follow the upgrade paths in our code. Here are couple of suggestion to make it simpler. This issue is meant capture them and we can spawn other tickets for each of the things we decide to do.
- We rely both on Lucene versions and Elasticsearch versions in our code. I think this is confusing. We should just rely on the Elasticsearch version.
- We capture the current status in index settings but we don't have clear separation in our naming between data upgrades (i.e., Lucene segments) and metadata upgrades. We have the following settings:
  - `index.version.upgraded` -> meta data was upgraded 
  - `index.version.minimum_compatible` -> lucene segements were upgraded (and uses lucene versions)
    they should be renamed to  `index.version.metadata_upgrade` & `index.version.data_upgrade` (or similar)
- For closed indices we ignore the minimum data requirement (see MetaDataIndexUpgradeService#checkSupportedVersion) as we aren't intending to go and read the lucene files. However, we _do_ upgrade the metadata files. This puts us in place where people can't downgrade back to a previous version, open the index and upgrade it. I _think_ that we we have closed indices which can't be opened we should refuse to start. The alternative is to leave the metadata alone - but then we can't guarantee that the Elasticsearch code of the new version can deal with it.
- The upgrade API sends upgrade commands to all the shards and when they respond successfully,we send a command to the master to mark the index as (data) upgraded. It would be nice to send the allocation IDs of the physical copies that were upgraded so the master can decide wether these are indeed contain all the _current_ primaries.

Other confusions & minor things:
- we have MetaDataUpdateSettingsService#upgradeIndexSettings which marks the index after upgrading the shard. We have MetaDataIndexUpgradeService which does all the metadata upgrading. We should move upgradeIndexSettings() into  MetaDataIndexUpgradeService
- MetaDataUpdateSettingsService#upgradeIndexSettings (or it's equivalent) should stay away from the meta data upgrade settings (it's generally OK because the cluster state was already upgraded at that point)
- MetaDataIndexUpgradeService#checkSupportedVersion ignores closed indices. That should be made explicit (either by the name or a parameter).

**Comment by jasontedor** [comment]
No interest in two-and-a-half years, closing. If any of these issues are serious and merit immediate attention, we should open a new issue targeting them.
